app/src/main/kotlin/compiler/common/dfa/CompositeDfa.kt

```diff
@@ -0,0 +1,32 @@
+package compiler.common.dfa
+
+import compiler.common.dfa.state_dfa.Dfa
+import compiler.common.dfa.state_dfa.DfaState
+
+class CompositeDfa<S : Comparable<S>, R, IR> (private val components: List<Pair<Dfa<S, IR>, R>>) : Dfa<S, R> {
```

Comment:
Currently we ignore the result type for composed dfas - I guess we can change `IR` to `Unit` (suggested) or modify `components` type to `List<Triple<Dfa<S, IR>, IR, R>>` (probably we don't need such functionality)

Reply:
I changed `IR` to `*` (so it's basically your first suggestion but hardcoding `Unit` when it's not necessary seems little weird to me).